pkg/container/versioned/value.go

```diff
@@ -0,0 +1,337 @@
+// Copyright Authors of Cilium
+
+package versioned
+
+import (
+	"math"
+	"slices"
+	"sync/atomic"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type Version uint64
+
+const (
+	// invalidVersion is never found from the value range.
+	// Also used as the upper bound for new values, so that the value is
+	// found when looking with 'maxVersion'
+	invalidVersion = Version(math.MaxUint64)
+
+	// maxVersion in a handle finds all non-removed values
+	maxVersion = Version(math.MaxUint64 - 1)
+
+	maxEmptyHandles = 1024
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "versioned")
+)
+
+// Handle is the object needed for getting values containing the latest version available at the
+// time of getting the Handle.
+type Handle struct {
+	// Handle stores version + 1 so that the zero value translates to invalidVersion
+	versionPlus1 Version
+}
+
+func newHandle(version Version) Handle {
+	return Handle{versionPlus1: Version(version) + 1}
+}
+
+// AllHandle returns a handle that matches all current/non-removed values
+func AllHandle() Handle {
+	return newHandle(maxVersion)
+}
+
+func (h Handle) getVersion() Version {
+	return h.versionPlus1 - 1
+}
+
+func (h Handle) IsValid() bool {
+	return h.versionPlus1 != 0
+}
+
+type atomicVersion struct {
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) Load() Version {
+	return Version(a.version.Load())
+}
+
+func (a *atomicVersion) Store(version Version) {
+	a.version.Store(uint64(version))
+}
+
+func (a *atomicVersion) CompareAndSwap(old, new Version) bool {
+	return a.version.CompareAndSwap(uint64(old), uint64(new))
+}
+
+// Manager manages the versioning for multiple VersionedValues.
+// It is to be stored wherever VersionedValues are stored
+type Manager struct {
+	version atomicVersion
+
+	// mutex protects 'handles'
+	mutex lock.RWMutex
+	// handles is an ordered list of outstanding handle versions. Outdated values can be removed
+	// when there are no outstanding handles for it.
+	handles []Version
+
+	// latest cleaned version
+	cleanedVersion Version
+
+	// cleaner is called with the earliest version that must be kept
+	// cleaner is called while 'mutex' is held!
+	cleaner func(keepVersion Version)
+}
+
+func NewVersionManager(cleaner func(keepVersion Version)) *Manager {
+	return &Manager{
+		cleaner: cleaner,
+	}
+}
+
+func (v *Manager) GetVersion() (current, next Version) {
+	current = v.version.Load()
+	return current, current + 1
+}
+
+func (v *Manager) PublishVersion(current, new Version) bool {
+	return v.version.CompareAndSwap(current, new)
+}
+
+type Closer func() bool
+
+// GetHandle returns a Handle for the current version and a closer for releasing old versions held
+// for this version.
+// 'name' is for debugging purposes only
+func (v *Manager) GetHandle(name string) (handle Handle, closer Closer) {
```

Comment:
Thanks for the review, I've changed this since your review. I thought about conforming to `io.Closer`, but hesitated since this is no "IO"' as such. The method is now called `Release()`, but given your suggestion I'll rename it as `Close` instead.

Reply:
done!